### ISSUE TITLE
All web platform features requiring a secure context should be so marked in the BCD

### ISSUE THREAD

**Issue by jpmedley** [mdn/browser-compat-data#4696]
I recently discovered that [service workers BCD](https://github.com/mdn/browser-compat-data/blob/master/api/ServiceWorker.json) does not list the requirements for secure context even though it is [required by the spec](https://w3c.github.io/ServiceWorker/#serviceworker-interface). Consequently this requirement does not appear on the interface's MDN page or in any of the other clients for this data set.

I'll fix this for service workers, but we should investigate this for all specs. I've generated a list of interfaces from Chrome's IDL. Our interface set is probably larger than other browsers but I don't assume it contains everything.

- [ ] AbsoluteOrientationSensor
- [ ] Accelerometer
- [ ] AmbientLightSensor
- [ ] ApplicationCache
- [ ] ApplicationCacheErrorEvent
- [ ] AudioWorklet
- [ ] AuthenticatorAssertionResponse
- [ ] AuthenticatorAttestationResponse
- [ ] AuthenticatorResponse
- [ ] BarcodeDetector
- [ ] Bluetooth
- [ ] BluetoothAdvertisingEvent
- [ ] BluetoothCharacteristicProperties
- [ ] BluetoothDevice
- [ ] BluetoothLEScan
- [ ] BluetoothManufacturerDataMap
- [ ] BluetoothRemoteGATTCharacteristic
- [ ] BluetoothRemoteGATTDescriptor
- [ ] BluetoothRemoteGATTServer
- [ ] BluetoothRemoteGATTService
- [ ] BluetoothServiceDataMap
- [ ] Cache
- [x] CacheStorage
- [ ] Clipboard
- [ ] ContactsManager
- [ ] CookieChangeEvent
- [ ] CookieStore
- [ ] Credential
- [ ] CredentialsContainer
- [ ] DetectedBarcode
- [ ] DetectedFace
- [ ] DetectedText
- [ ] DeviceMotionEvent
- [ ] DeviceMotionEventAcceleration
- [ ] DeviceMotionEventRotationRate
- [ ] DeviceOrientationEvent
- [ ] FaceDetector
- [ ] FederatedCredential
- [ ] FileSystemDirectoryHandle
- [ ] FileSystemFileHandle
- [ ] FileSystemHandle
- [ ] FileSystemWriter
- [ ] Gyroscope
- [ ] HID
- [ ] HIDCollectionInfo
- [ ] HIDConnectionEvent
- [ ] HIDDevice
- [ ] HIDInputReportEvent
- [ ] HIDReportInfo
- [ ] HIDReportItem
- [ ] IdleDetector
- [ ] IdleState
- [ ] Keyboard
- [ ] KeyboardLayoutMap
- [ ] LinearAccelerationSensor
- [ ] Lock
- [ ] LockManager
- [ ] MIDIAccess
- [ ] MIDIConnectionEvent
- [ ] MIDIInput
- [ ] MIDIInputMap
- [ ] MIDIMessageEvent
- [ ] MIDIOutput
- [ ] MIDIOutputMap
- [ ] MIDIPort
- [ ] Magnetometer
- [ ] MediaDeviceInfo
- [ ] MediaDevices
- [ ] MediaKeyMessageEvent
- [ ] MediaKeySession
- [ ] MediaKeyStatusMap
- [ ] MediaKeySystemAccess
- [ ] MediaKeys
- [ ] MerchantValidationEvent
- [x] NDEFMessage
- [x] NDEFRecord
- [ ] NFCErrorEvent
- [ ] NFCReader
- [ ] NFCReadingEvent
- [ ] NFCWriter
- [ ] NavigationPreloadManager
- [ ] OrientationSensor
- [ ] PasswordCredential
- [ ] PaymentAddress
- [ ] PaymentMethodChangeEvent
- [ ] PaymentRequest
- [ ] PaymentResponse
- [ ] Presentation
- [ ] PresentationAvailability
- [ ] PresentationConnection
- [ ] PresentationConnectionAvailableEvent
- [ ] PresentationConnectionCloseEvent
- [ ] PresentationConnectionList
- [ ] PresentationReceiver
- [x] PresentationRequest
- [ ] PublicKeyCredential
- [ ] RTCIceTransport
- [ ] RTCQuicStream
- [ ] RTCQuicTransport
- [ ] RelativeOrientationSensor
- [ ] SMS
- [ ] SMSReceiver
- [ ] Sensor
- [ ] SensorErrorEvent
- [ ] Serial
- [ ] SerialPort
- [ ] ServiceWorker
- [ ] ServiceWorkerContainer
- [ ] ServiceWorkerRegistration
- [ ] StorageManager
- [ ] TextDetector
- [ ] USB
- [ ] USBAlternateInterface
- [ ] USBConfiguration
- [ ] USBConnectionEvent
- [ ] USBDevice
- [ ] USBEndpoint
- [ ] USBInTransferResult
- [ ] USBInterface
- [ ] USBIsochronousInTransferPacket
- [ ] USBIsochronousInTransferResult
- [ ] USBIsochronousOutTransferPacket
- [ ] USBIsochronousOutTransferResult
- [ ] USBOutTransferResult
- [ ] WakeLock
- [ ] Window
- [ ] WorkerNavigator
- [ ] Worklet
- [ ] XR
- [ ] XRAnchor
- [ ] XRAnchorSet
- [ ] XRBoundedReferenceSpace
- [ ] XRFrame
- [ ] XRHitResult
- [ ] XRHitTestOptions
- [ ] XRHitTestResult
- [ ] XRHitTestSource
- [ ] XRInputSource
- [ ] XRInputSourceArray
- [ ] XRInputSourceEvent
- [ ] XRInputSourcesChangeEvent
- [ ] XRPlane
- [ ] XRPlaneDetectionState
- [ ] XRPlaneSet
- [ ] XRPose
- [ ] XRRay
- [ ] XRReferenceSpace
- [ ] XRReferenceSpaceEvent
- [ ] XRRenderState
- [ ] XRRigidTransform
- [ ] XRSession
- [ ] XRSessionEvent
- [ ] XRSpace
- [ ] XRView
- [ ] XRViewerPose
- [ ] XRViewport
- [ ] XRWebGLLayer
- [ ] XRWorldInformation
- [ ] XRWorldTrackingState

**Comment by Elchi3** [comment]
The data guideline for this is https://github.com/mdn/browser-compat-data/blob/master/docs/data-guidelines.md#secure-context-required-secure_context_required

If you could provide a list of interfaces that would need this then this issue would become actionable. Thank you!

**Comment by Elchi3** [comment]
Thanks Joe! Looking at the list, do you you think it is useful to note down secure context compat data even when the interface has never been exposed without secure context? I believe this is the case for new APIs, like XR interfaces.

I think this might only be interesting if there is a change to "now requires secure context in version x" which differs to "since version y it was possible to use this in any context". The list would be smaller than I guess.

**Comment by jpmedley** [comment]
I had the same question. I think it needs to be noted in every case for the next few years at least. According to the [http archive](https://httparchive.org/reports/state-of-the-web?start=2019_01_01&end=latest&view=list#pctHttps) 20% of web traffic still uses https. That's an awful lot of people who still need to know whether a particular feature is usable on their site. 

Perhaps it could be presented on MDN in some way other than a separate line in the compat tables?

**Comment by Elchi3** [comment]
> Perhaps it could be presented on MDN in some way other than a separate line in the compat tables?

I think that interfaces that are exposed to https from the start of their existence, we can note that on the MDN page but don't need it in the compat tables.

For interfaces where it was available everywhere first but then in later versions changed availability to only https, we should have the compat table row.

**Comment by jpmedley** [comment]
> I think that interfaces that are exposed to https only from the start of their existence, we can note that on the MDN page but don't need it in the compat tables.

Which would mean that information isn't available to [other projects that use the data](https://github.com/mdn/browser-compat-data#projects-using-the-data).

**Comment by Elchi3** [comment]
That's true! If we hear about interest from other projects, I can justify spending time on this to have this for all interfaces or what is needed to meet their requirements. Right now, it seems we would serve MDN only and could go with a less time consuming way by only adding it to interfaces where the availability changed from all contexts to https context. Are you are aware of anyone needing to deal with this info besides MDN docs?

**Comment by ExE-Boss** [comment]
> - [ ] [`CSS`](https://developer.mozilla.org/docs/Web/API/CSS)

That’s… surprising.

This probably only applies to Houdini Worklets and not to [`CSS.escape(…)`](https://developer.mozilla.org/docs/Web/API/CSS/escape) or [`CSS.supports(…)`](https://developer.mozilla.org/docs/Web/API/CSS/supports) (the latter being the programmatic version of the [`@supports` at‑rule](https://developer.mozilla.org/docs/Web/CSS/@supports)).

**Comment by jpmedley** [comment]
I dug into the IDL. It looks like it's just CSS.paintWorklet (which makes sense to me). It made the list of interfaces because of an IDL quirk my script doesn't account for. I've removed it.

**Comment by bershanskiy** [comment]
Some features are exposed only to secure contexts, but their methods and attributes are not explicitly marked in WebIDL for compatibility.
For example, Battery Status API its primary entry point, [`Navigator.getBattery()`](https://wiki.developer.mozilla.org/en-US/docs/Web/API/Navigator/getBattery)
 - in modern Chrome, is available in insecure (and third-party) contexts to maintain backward compatibility, but just never resolves to preserve security.
 - [specked out](https://w3c.github.io/battery/#ref-for-dom-navigator-getbattery-1) to throw `SecurityError` in insecure contexts

My question is, should BCD just mark such interfaces with `secure_context_required` or is there a more specific format?
